aten/src/ATen/core/ivalue.cpp

```diff
@@ -132,6 +153,101 @@ bool IValue::overlaps(const IValue& rhs) const {
+  // not a boolean. Mimic that behavior here.
+  TORCH_INTERNAL_ASSERT(eq.isTensor());
+  return eq.toTensor().is_nonzero();
+}
+
+IValue IValue::equals(const IValue& rhs) const {
+  const IValue& lhs = *this;
+  switch (lhs.tag) {
+    case Tag::None:
+      // In Python you're not supposed to do this comparison apparently. Not
+      // sure if we should warn here or what
+      return rhs.isNone();
+    case Tag::Tensor:
+      if (!rhs.isTensor()){return false;}
+      return lhs.toTensor().eq(rhs.toTensor());
+    case Tag::Double:
+      return rhs.isDouble() && lhs.toDouble() == rhs.toDouble();
+    case Tag::Int:
+      return rhs.isInt() && lhs.toInt() == rhs.toInt();
+    case Tag::Bool:
+      return rhs.isBool() && lhs.toBool() == rhs.toBool();
+    case Tag::String:
+      return rhs.isString() && lhs.toStringRef() == rhs.toStringRef();
+    case Tag::GenericDict:
+      return rhs.isGenericDict() && lhs.toGenericDict() == rhs.toGenericDict();
+    case Tag::Tuple:
+      return rhs.isTuple() && lhs.toTuple() == rhs.toTuple();
+    case Tag::Device:
+      return rhs.isDevice() && lhs.toDevice() == rhs.toDevice();
+    case Tag::GenericList:
+      return rhs.isList() && lhs.toList() == rhs.toList();
+    case Tag::Blob:
+      return rhs.isBlob() && lhs.isAliasOf(rhs);
+    case Tag::Future:
+      return rhs.isFuture() && lhs.isAliasOf(rhs);
+    case Tag::RRef:
+      return rhs.isRRef() && lhs.isAliasOf(rhs);
+    case Tag::Object:
+      return rhs.isObject() && lhs.isAliasOf(rhs);
+    case Tag::PyObject:
+      return rhs.isPyObject() && lhs.isAliasOf(rhs);
+    case Tag::Capsule:
+      return rhs.isCapsule() && lhs.isAliasOf(rhs);
+    case Tag::Uninitialized:
+      TORCH_INTERNAL_ASSERT(
+          false,
+          "Uninitialized IValues are internal-only,",
+          "they should never participate in equality comparison");
+  }
+  // the above switch should be exhaustive
+  TORCH_INTERNAL_ASSERT(false, "we should never reach here")
+}
+
+
+static bool isUndefinedTensor(const IValue& iv) {
+  return iv.isTensor() && !iv.toTensor().defined();
+}
+bool IValue::is(const IValue& rhs) const {
+  const IValue& lhs = *this;
+  // Special handling for undefined tensors:
+  // 1. Undefined_tensor is None and vice versa.
+  if ((isUndefinedTensor(lhs) && rhs.isNone()) ||
+      (lhs.isNone() && isUndefinedTensor(rhs))) {
+    return true;
+  }
+  // 2. Undefined_tensor is Undefined_tensor.
+  if (isUndefinedTensor(lhs) && isUndefinedTensor(rhs)) {
+    return true;
+  }
+
+  if (lhs.isTensor()) {
+    // Compare the underlying tensor implementations, as we may have two
+    // distinct `at::Tensor` objects pointing to the same impl.
+    return rhs.isTensor() && lhs.toTensor().is_same(rhs.toTensor());
+  }
+
+  if (lhs.is_intrusive_ptr) {
+    TORCH_INTERNAL_ASSERT(rhs.is_intrusive_ptr);
+    return lhs.payload.as_intrusive_ptr == rhs.payload.as_intrusive_ptr;
+  }
+  // TODO should we do this?
```

Comment:
This is going to stack overflow?

Reply:
the implementation of equality doesn't call into `is`. Nevertheless, maybe we should just return false here…once we get to this point it's not really a valid comparison anymore.